### ISSUE TITLE
linter: `unicorn/consistent-function-scoping` false positive when an inner function references a generic type of outer function

### ISSUE THREAD

**Issue by jz-cp** [oxc-project/oxc#24336]
### What version of Oxlint are you using?

latest

### What command did you run?

oxlint . --deny-warnings --type-aware --type-check --report-unused-disable-directives

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
// Replace this with your actual config file contents.
// It can be helpful to keep it minimal, if possible, but please ensure that it actually reproduces the issue you're reporting.
{
  "categories": {
    "correctness": "deny",
    "nursery": "deny",
    "pedantic": "deny",
    "perf": "deny",
    "restriction": "deny",
    "style": "deny",
    "suspicious": "deny",
  },
  "plugins": ["unicorn"],
  "rules": {
    "unicorn/consistent-function-scoping": "error"
  }
}
```

### What happened?

**What the code does:**
Outer function defines a generic param `T`. Inner function accepts a param of type `T`.

**What I expected:**
No errors

**What happened:**
```
Oxc Error: Function `inner` does not capture any variables from its parent scope
```

**Code**
```ts
function outer<T>() {
  function inner(param1: T) {
    console.log(param1)
  }
}
```

**Comment by camc314** [comment]
This seems like a niche case. Do you have a real world example of this code.

Why not just move it out of outer, and remove the type param?

**Comment by jz-cp** [comment]
This is obviously a trimmed down version to replicate and demonstrate the issue. In my real code, the outer and inner functions are doing much more. And the inner function must be defined inside of the outer one, since the outer one uses some inference to determine the shape of the generic T.

**Comment by jz-cp** [comment]
Here's a playground link
https://playground.oxc.rs/?lintRules=unicorn%2Fconsistent-function-scoping&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=function+outer%3CT%3E%28%29+%7B%0A++function+inner%28param1%3A+T%29+%7B%0A++++console.log%28param1%29%0A++%7D%0A%7D%0A

**Comment by camc314** [comment]
I get the concept of why you'd want to do it, but in practice wouldn't it just be better to disable it for that line?

One of the goals of the rule is to help improve performance, as a new fn instance will be created for every time the outer is called. Keeping the function contained within the outer just to satisfy the type system feels like an antipattern.

**Comment by durationextender** [comment]
is this issue still open? if yes i would like to work on it @camc314

**Comment by camc314** [comment]
> is this issue still open? if yes i would like to work on it @camc314

It's still open, but I need a little bit more context before deciding if it's a bug vs intended behaviour.

**Comment by jz-cp** [comment]
@camc314 What else can I add to this?

**Comment by lukpsaxo** [comment]
I hit this on our project - we have a react component - Menu - which takes a menu key type param. The menu needs to pass a function to a sub react component and it is not possible to use inference - in this case we need to explicitly refer to the type parameter.
So we are stuck either casting the function into the specific type we need based on the type param, or disabling the eslint rule error and saying "invalid error on this line".

The original unicorn plugin does not report this as a error.
